processor/maskprocessor/processor.go

```diff
@@ -0,0 +1,187 @@
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+const (
+	resourceField   = "resource"
+	attributesField = "attributes"
+	bodyField       = "body"
+)
+
+// processor is the processor used to mask data.
+type processor struct {
+	logger           *zap.Logger
+	cfg              *Config
+	rules            map[string]*regexp.Regexp
+	maskResourceFunc func(k string, v pcommon.Value) bool
+	maskAttrsFunc    func(k string, v pcommon.Value) bool
+}
+
+// newProcessor creates a new mask processor.
+func newProcessor(logger *zap.Logger, cfg *Config) *processor {
+	return &processor{
+		logger: logger,
+		cfg:    cfg,
+	}
+}
+
+// start is used to start the processor.
+func (p *processor) start(context.Context, component.Host) error {
+	rules, err := p.cfg.CompileRules()
+	if err != nil {
+		return err
+	}
+
+	p.rules = rules
+	p.maskResourceFunc = p.createMaskFunc(resourceField)
+	p.maskAttrsFunc = p.createMaskFunc(attributesField)
+	return nil
+}
+
+// processLogs masks incoming logs.
+func (p *processor) processLogs(_ context.Context, ld plog.Logs) (plog.Logs, error) {
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resource := ld.ResourceLogs().At(i)
+		resource.Resource().Attributes().Range(p.maskResourceFunc)
+		for j := 0; j < resource.ScopeLogs().Len(); j++ {
+			scope := resource.ScopeLogs().At(j)
+			for k := 0; k < scope.LogRecords().Len(); k++ {
+				logs := scope.LogRecords().At(k)
+				logs.Attributes().Range(p.maskAttrsFunc)
+				p.maskValue(bodyField, logs.Body())
+			}
+		}
+	}
+
+	return ld, nil
+}
+
+// processTraces masks incoming traces.
+func (p *processor) processTraces(_ context.Context, td ptrace.Traces) (ptrace.Traces, error) {
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		resource := td.ResourceSpans().At(i)
+		resource.Resource().Attributes().Range(p.maskResourceFunc)
+		for j := 0; j < resource.ScopeSpans().Len(); j++ {
+			scope := resource.ScopeSpans().At(j)
+			for k := 0; k < scope.Spans().Len(); k++ {
+				spans := scope.Spans().At(k)
+				spans.Attributes().Range(p.maskAttrsFunc)
+			}
+		}
+	}
+
+	return td, nil
+}
+
+// processMetrics masks incoming metrics.
+func (p *processor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	for i := 0; i < md.ResourceMetrics().Len(); i++ {
+		resource := md.ResourceMetrics().At(i)
+		resource.Resource().Attributes().Range(p.maskResourceFunc)
+		for j := 0; j < resource.ScopeMetrics().Len(); j++ {
+			scope := resource.ScopeMetrics().At(j)
+			for k := 0; k < scope.Metrics().Len(); k++ {
+				metrics := scope.Metrics().At(k)
+				switch metrics.Type() {
+				case pmetric.MetricTypeSum:
+					p.processSum(metrics.Sum())
+				case pmetric.MetricTypeGauge:
+					p.processGauge(metrics.Gauge())
+				case pmetric.MetricTypeSummary:
+					p.processSummary(metrics.Summary())
+				}
```

Comment:
Any reason we don't support histogram & exponential histograms here?

Reply:
@jmwilliams89 any thoughts on this? @BinaryFissionGames are those still "legacy" types? I thought at one point there was discussion of only using them if you absolutely have to. If that's still the case I think it's ok to skip them. This processor likely isn't going to be targeted at metrics as much as logs/traces.

Reply:
@cpheps I think it's OK to just ignore ExponentialHistogram, but maybe we should note in our docs that it's not supported?

Reply:
@BinaryFissionGames decided to just add the ExponentialHistogram type.